### ISSUE TITLE
yubikey not locked after removal (CAPI)

### ISSUE THREAD

**Issue by DonEstefan** [NoMoreFood/putty-cac#141]
When removing an unlocked yubikey from the USB-Slot, I'd expect the key to lock itself and ask for another PIN verification before next use. With putty-cac this does not happen. Yubikey is still unlocked after re-insertion.  I have to restart putty-cac in order to get a new PIN unlock prompt. 

- Putty-cac 0.81 and "Force PIN Caching" is disabled
- Windows 11
- Yubikey Firmware 5.7.1
- YubiKeySmartCardMinidriver 4.6.3.252

Is this a bug in putty-cac? Maybe some strange [CAPI session](https://docs.yubico.com/yesdk/yubikey-api/Yubico.YubiKey.Piv.PivSession.html) handling?

Steps to reproduce;
  1) configure yubikey for CAPI authentication
  2) add yubikey to pageant via CAPI
  3) start ssh session and authenticate via yubikey. Note that yubikey asks for PIN unlock on first use.
  4) close ssh session (but not pageant) 
  5) remove yubikey from USB
  6) re-instert yubikey
  8) start the same ssh session from step 3. Note that yubikey authentication works fine _without_ PIN verification prompt being shown.

**Comment by NoMoreFood** [comment]
Can you provide a screenshot of the PIN prompt you receive when you _are_ initially prompted? I just want to make sure it absolutely isn't going through PuTTY CACs PIN caching code.  I believe the behavior you are seeing is an artifact of the caching that can occur at the minidriver / OS level that's usually per-process based.  It's possible this is triggered by an unintentional leftover CAPI handle on the PuTTY CAC side, but we've been pretty good about handle tracking.

**Comment by NoMoreFood** [comment]
Yeah, so that's definitely the Windows-controlled PIN dialog box (i.e. out of PuTTY CAC's control).  I suspect what you found explains the behavior you are seeing.  I cannot find a way to override this behavior generically.  I agree it really _ought_ to invalidate the cache when the key is removed.  It seems like under Windows 10, that would be the responsibility of the Yubikey minidriver.

**Comment by NoMoreFood** [comment]
> They suggest there might be a timeout limit in the agent

If by "agent" they mean Pageant, this is certainly not the case.  I'll see if I can reproduce this behavior with a smaller test case.  In Pageant, we really don't do anything directly with the card; we interact with the certificate APIs which, in turn, communicate with the CSP and minidriver.

**Comment by bluikko** [comment]
I've never seen this with Pageant so I obtained 1 YubiKey to test this particular report.  
PIN is correctly asked after unplugging+replugging YubiKey. I used the old drivers that were discontinued some years ago.

Suspect it's a configuration issue or other user issue and not something with PuTTy-CAC.

**Comment by bluikko** [comment]
> > I used the old drivers that were discontinued some years ago
> 
> @bluikko, what exactly do you mean by "old drivers"? The yubikey smartcard driver I'm using is still seeing regular updates (see [here](https://support.yubico.com/hc/en-us/articles/14400158281756-YubiKey-Smart-Card-Minidriver-release-notes)).

There have been historically 2 different minidrivers for the YubiKeys. I tested with the driver that was discontinued a few years (maybe something like 3-5 years?) ago and replaced by the "current" driver. Sorry I don't have more details, this was just a test on an available system and I have not followed what exactly is going on with YubiKeys since a few years.
